lxd/api_cluster.go

```diff
@@ -1212,13 +1210,31 @@ func rebalanceMemberRoles(d *Daemon) error {
 		return nil
 	}
 
+	// Process demotions of offline nodes immediately.
+	for _, node := range nodes {
+		if node.Address != address || node.Role != db.RaftSpare {
+			continue
+		}
+
+		if cluster.HasConnectivity(d.endpoints.NetworkCert(), address) {
+			break
+		}
+
+		err := d.gateway.DemoteOfflineNode(node.ID)
+		if err != nil {
+			return errors.Wrapf(err, "Demote offline node %s", node.Address)
+		}
+
+		goto again
+	}
+
 	// Tell the node to promote itself.
 	err = changeMemberRole(d, address, nodes)
 	if err != nil {
 		return err
 	}
 
-	return nil
+	goto again
```

Comment:
So we can only get out of this function if address is empty?

Reply:
Yes. Essentially `address` won't be empty until there are no more roles changes that should be performed.

Reply:
Ok